python/lsst/butler/config.py

```diff
@@ -0,0 +1,320 @@
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <https://www.lsstcorp.org/LegalNotices/>.
+#
+
+import collections
+import copy
+import os
+import sys
+import warnings
+import yaml
+
+import lsst.utils
+
+from yaml.representer import Representer
+yaml.add_representer(collections.defaultdict, Representer.represent_dict)
+
+# UserDict and yaml have defined metaclasses and Python 3 does not allow multiple
+# inheritance of classes with distinct metaclasses. We therefore have to
+# create a new baseclass that Config can inherit from. This is because the metaclass
+# syntax differs between versions
+
+class _ConfigMeta(type(collections.UserDict), type(yaml.YAMLObject)):
+    pass
+
+
+class _ConfigBase(collections.UserDict, yaml.YAMLObject, metaclass=_ConfigMeta):
+    pass
+
+
+class Config(_ConfigBase):
+    """Config implements a datatype that is used by Butler for configuration parameters.
+    It is essentially a dict with key/value pairs, including nested dicts (as values). In fact, it can be
+    initialized with a dict. The only caveat is that keys may NOT contain dots ('.'). This is explained next:
+    Config extends the dict api so that hierarchical values may be accessed with dot-delimited notiation.
+    That is, foo.getValue('a.b.c') is the same as foo['a']['b']['c'] is the same as foo['a.b.c'], and either
+    of these syntaxes may be used.
+
+    Storage formats supported:
+    - yaml: read and write is supported.
+    """
+
+    def __init__(self, other=None):
+        """Initialize the Config. Other can be used to initialize the Config in a variety of ways:
+        other (string) Treated as a path to a config file on disk. Must end with '.paf' or '.yaml'.
+        other (Config) Copies the other Config's values into this one.
+        other (dict) Copies the values from the dict into this Config.
+        """
+        collections.UserDict.__init__(self)
+
+        if other is None:
+            return
+
+        if isinstance(other, collections.Mapping):
+            self.update(other)
+        elif isinstance(other, Config):
+            self.data = copy.deepcopy(other.data)
+        elif isinstance(other, str):
+            # if other is a string, assume it is a file path.
+            self.__initFromFile(other)
+        else:
+            # if the config specified by other could not be loaded raise a runtime error.
```

Comment:
This comment is not strictly correct. This error is raised if other was not recognized, not if it could not be loaded.